apps/web/vibes/soul/sections/product-detail/product-detail-form.tsx

```diff
@@ -72,10 +76,18 @@ export function ProductDetailForm<F extends Field>({
     }),
     { quantity: 1 },
   );
+
   const [{ lastResult }, formAction] = useActionState(action, {
     fields,
     lastResult: null,
   });
+
+  useEffect(() => {
+    if (lastResult?.status === 'success') {
+      toast.success(successMessage);
+    }
+  }, [lastResult, successMessage]);
```

Comment:
We should do this `onSubmit`, not on `useEffect`

Reply:
You're right! 🤦

Reply:
Question: what is the best way?

`onSubmit` in `useForm` is called before server validation.

Reply:
That's a good question. What about something like this?

```tsx
function SomeForm({ action }) {
  const [{ lastResult }, formAction, isPending] = useActionState(action, { lastResult: null })
  const [form] = useForm({
    lastResult,
    onSubmit(event, formData) {
      // prevent `formAction` from being called, we'll call `action` ourselves
      event.preventDefault()

      startTransition(async () => {
        const payload = await action(lastResult, formData)

        await toaster.success(payload.successMessage)
      })
    }
  })

  return <form action={formAction} onSubmit={form.onSubmit} />
}
```

This has problems with keeping the action state up to date since we're not calling `formAction`. Ideally we can call `formAction` and await that but I'm not sure that's possible.

We could also pass a client action to `useActionState` and that would work well. It would mean we don't support progressive enhancement, but that's probably fine.

```tsx
function SomeForm({ action }) {
  const [{ lastResult }, formAction, isPending] = useActionState(async (prevState, formData) => {
    const nextState = await action(lastResult, formData)

    await toaster.success(nextState.successMessage)

    return nextState
  }, { lastResult: null })
  const [form] = useForm({ lastResult })

  return <form action={formAction} onSubmit={form.onSubmit} />
}
```

Reply:
Maybe let's just continue using `useEffect` for now?

Reply:
I implemented a solution that works but has a smell, let me know what you think.

Reply:
Reverted to `useEffect`.

Reply:
I think using `useEffect` is fine for now